extensions/bustache/ApplyTemplate.h

```diff
@@ -61,7 +61,7 @@ class ApplyTemplate : public core::Processor {
   class WriteCallback : public OutputStreamCallback {
    public:
     WriteCallback(const std::string &templateFile, const std::shared_ptr<core::FlowFile> &flow_file);
-    int64_t process(std::shared_ptr<io::BaseStream> stream);
+    int64_t process(const std::shared_ptr<io::BaseStream>& stream);
```

Comment:
I would say, that using `const shared_ptr<>&` is certainly an improvement, as @lordgamez said it does not force us to take ownership, and it is agreement with the proliferation of managed pointers in the rest of the system (so it lets us call into interfaces expecting a managed ptr), I don't have a strong opinion about it though, because I plan on removing the whole `*StreamCallback` stuff, so if you feel like it should be changed, decide which one should it be:

1. shared_ptr<BaseStream>
2. const shared_ptr<BaseStream>&
3. BaseStream&
4. BaseStream*
5. gsl::not_null<BaseStream&>

Reply:
Understood. I'd go for 3.

Reply:
I'm okay with any options of 2-5.

Reply:
Note: I didn't really review this PR, just briefly looked at the diff and shared my thoughts about it. Me generally preferring 3. is because it's normally not nullable and doesn't enforce any lifetime semantics of the object, which is normally the ideal when we only care about the object. When storing as a data member, I prefer 5. or 4, because they are mutable and sometimes one needs nullable members to be able to implement move semantics.

With that said, I can accept any other way, in any case, especially given good reasoning (e.g. the proliferation of shared_ptr throughout the codebase).

Reply:
as it turns out there are a number of implementations that store the stream, so that (third-party) callbacks may access it, e.g. `CompressContent::WriteCallback`, `JniByteInputStream`, `FocusArchiveEntry::ReadCallback`, `ArchiveMerge::WriteCallback`, `UnfocusArchiveEntry::WriteCallback`, `LuaInputStreamCallback`, `PyOutputStreamCallback`, moreover the hash functions in `HashContent` expect a `shared_ptr`

base on this coupled with the diff explosion that would result from rewriting all the `stream->write` to `stream.write`, I propose we leave this be for now, and return to it in a later PR